livedata/src/main/java/io/getstream/chat/android/livedata/EventBatchUpdate.kt

```diff
@@ -0,0 +1,119 @@
+package io.getstream.chat.android.livedata
+
+import io.getstream.chat.android.client.models.Channel
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.livedata.entity.ChannelEntity
+import io.getstream.chat.android.livedata.entity.MessageEntity
+import io.getstream.chat.android.livedata.extensions.users
+
+/**
+ * EventBatchUpdate helps you efficiently implement a 4 step batch update process
+ * It updates multiple messages, users and channels at once.
+ *
+ * val batch = EventBatchUpdate(domainImpl)
+ *
+ * as a first step specify which channels and messages to fetch
+ * batch.addToFetchChannels()
+ * batch.addToFetchMessages()
+ *
+ * as a second step, load the required data for batch updating using
+ * batch.fetch()
+ *
+ * third, add the required updates via
+ * addUser, addChannel and addMessage methods
+ *
+ * fourth, execute the batch using
+ * batch.execute()
+ */
+internal class EventBatchUpdate(private val domainImpl: ChatDomainImpl) {
```

Comment:
The instructions above makes me feel that some kind of a builder pattern would be more suitable here instead. 

All the `add...()` methods could be provided by `BatchUpdate.Builder` and `fetch()` would be done in `BatchUpdate.Builder.build()`.

Eventually, this would allow the instance of `BatchUpdate` to have immutable, private properties only, making the `BatchUpdate.execute()` operation thread-safe. 

WDUT?

Reply:
```kotlin
internal class EventBatchUpdate.Builder(private val domain: ChatDomain) {
  fun add...()
  fun add...()

  fun build() {
    //fetch all the required data here and create EventBatchUpdate
    return EventBatchUpdate() // EventBatchUpdate would be an immutable class
  } 
} 
```
```kotlin
internal class EventBatchUpdate(/*only immutable props here*/) {
  fun execute()
}
```

I'm thinking about something like above.

Reply:
pro: type system prevents you from using it the wrong way
con: it's a bit more complex

In most languages the simple solution would be preferred. Though in the kotlin ecosystem I think the builder pattern looks more natural. Going to do that, fits more with the ecosystem.